### PR TITLE
PE-796 Create Gluu Laravel Guard

### PR DESCRIPTION
**Pull request by buiminhducdiag**
Ticket:
https://diag.atlassian.net/browse/PE-796

### CODE REVIEW

**Review by toanauthe — CHANGES_REQUESTED** [review]
I think we should separating this to something like 
```
class GluuGuard
{
  public function check()
  public function user()
  public function setUser()
  public function validate()
}
```
Then when you register the guard with laravel, we can use something like this
Auth::check()..
-> Then Other services no need to re-implement the authentication again, they just define the guard to gluu-guard

**Comment by buiminhducdiag** [comment]
> I think we should separating this to something like
> 
> ```
> class GluuGuard
> {
>   public function check()
>   public function user()
>   public function setUser()
>   public function validate()
> }
> ```
> 
> Then when you register the guard with laravel, we can use something like this Auth::check().. -> Then Other services no need to re-implement the authentication again, they just define the guard to gluu-guard

File: src/App/GluuGuard.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        private AuthenticationInterface $auth
+    ){
+
+    }
+    public function validateToken()
+    {
+        $result = $this->auth->inspectToken($this->request->bearerToken());
+        if ($result) {
+            $userRepository = new UserRepository();
+            return $userRepository->allQuery(['username' => $result['username']])->first();
```
Comment (bradleynguyendiag) [inline src/App/GluuGuard.php:23]: Can we do some caching here. it should not call to Gluu server every time even token still valid.

Comment (buiminhducdiag) [reply src/App/GluuGuard.php:23]: Well, technically we can add caching here, but why we should do it? Validating token has its meaning when we can validate its validation with Gluu server. If we cache, we must deal with the cache itself, and how long it should be cached. And please think about security, if there is some reason, we need to remove the token from Gluu, and we still cache the token and its validation, then what happens?

Comment (bradleynguyendiag) [reply src/App/GluuGuard.php:23]: That case never happen.

File: src/Providers/GluuServiceProvider.php [hunk 2]
```diff
@@ -1,18 +1,32 @@
 <?php
 namespace Gluu\Providers;
 
+use App\Repositories\User\UserRepository;
 use Gluu\App\AuthenticationInterface;
 use Gluu\App\GluuClient;
+use Gluu\App\GluuGuard;
 use GuzzleHttp\Client;
```
Comment (toanauthe) [inline src/Providers/GluuServiceProvider.php:8]: use Illuminate\Support\Facades\Auth;

File: src/App/GluuGuard.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    ){
+
+    }
+    public function validateToken()
+    {
+        if (Cache::has($this->request->bearerToken())) {
+            $result = Cache::get($this->request->bearerToken());
+        } else {
+            $result = $this->auth->inspectToken($this->request->bearerToken());
+            Cache::put($this->request->bearerToken(), $result, now()->addHours(4));
```
Comment (bradleynguyendiag) [inline src/App/GluuGuard.php:25]: Can we move 4 to config?

File: src/Config/gluu.php [hunk 4]
```diff
@@ -12,4 +12,5 @@
     'revoke_token_uri' => env('REVOKE_TOKEN_URI', '/oxauth/restv1/revoke'),
     'update_user_uri' => env('UPDATE_USER_URI', ''),
     'default_email_domain_prefix' => env('GLUU_DEFAULT_EMAIL_DOMAIN_PREFIX', '@gluu.diag.vn'),
+    'gluu_guard_cache_hour' => 4,
```
Comment (bradleynguyendiag) [inline src/Config/gluu.php:15]: Pls set to use env

**Review by bradleynguyendiag — APPROVED** [review]

**Comment by buiminhducdiag** [comment]
> I think we should separating this to something like
> 
> ```
> class GluuGuard
> {
>   public function check()
>   public function user()
>   public function setUser()
>   public function validate()
> }
> ```
> 
> Then when you register the guard with laravel, we can use something like this Auth::check().. -> Then Other services no need to re-implement the authentication again, they just define the guard to gluu-guard

Done

**Comment by buiminhducdiag** [comment]
> I think we should separating this to something like
> 
> ```
> class GluuGuard
> {
>   public function check()
>   public function user()
>   public function setUser()
>   public function validate()
> }
> ```
> 
> Then when you register the guard with laravel, we can use something like this Auth::check().. -> Then Other services no need to re-implement the authentication again, they just define the guard to gluu-guard

Done

**Review by toanauthe — APPROVED** [review]

**Pull request merged**